### PR TITLE
Omni: add Windows Torch 2.13 CUTE and SolAttn

### PR DESCRIPTION
**Pull request by xiangyuT**
Add the current Windows BMG CUTE and SolAttn sidecar, Torch 2.13/oneAPI 2026 packaging support, ABI-suffixed extension discovery, and ComfyUI routing. Replace the Windows deployment and wheel guides with current-only ComfyUI 0.34.0, Kitchen 0.2.31, AIMDO 0.4.15, and provider installation contracts; remove obsolete milestone constraints and conclusions.

### CODE REVIEW

**Comment by yzwzhanghao2** [comment]
Thanks for landing the Windows Torch 2.13 CUTE/Sol-Attn path. I tested PR #659 head `affc2ea6` on a physical Arc B580 (G21) and hit a reproducible forward hang before the real ComfyUI gate.

Environment:
- Windows, Arc B580, PCI ID `0xE20B`, `intel_gpu_bmg_g21`
- driver `32.0.101.8974`
- Python 3.13.12
- PyTorch `2.13.0+xpu` (`torch.version.xpu == 20260000`)
- oneAPI compiler/ocloc 2026.1; oneDNN 3.11.2
- pinned sycl-tla `2fc09973bfdf15755090fcb0e3b6ad236408a992`

Because the PR's Windows CUTE target is hard-coded to `bmg-g31`, I made one isolated build-only change so `WINDOWS_CUTE_AOT_TARGETS["bmg"]` could be overridden, then built with `bmg-g21`. The wheel built and imported successfully:
- `omni_xpu_kernel 0.2.0b2+torch213.bmg`
- SHA256 `4E35336F68E15FCCFB13A90B4F1FBFD9975DA6C946F9969A485F17994B7F68EE`
- `cute.is_available()`, `supports_d128_bhld()`, `supports_minimax_h3_vae_d64()`, and `supports_sol_attn()` all returned true.

The matching PyTorch SDPA control completed with finite output, but each native CUTE forward remained in a CPU busy wait:
1. H3 QKV-backed BF16 `[1,56,31,128]`, stride `[7168,128,21504,1]`: >124 s; repeat >34 s.
2. Dense BF16 BHLD `[1,4,31,128]`: >34 s.
3. Legacy FP16 BLHD `[1,256,8,128]`: >34 s.

There was no exception or explicit `DEVICE_LOST`; the Python process used about 1.4 GiB RAM and had to be terminated. Fresh Torch XPU matmul/device checks passed after each termination. I stopped before Sol-Attn forward and the real H3 `[1,56,7461,128]` gate.

One possibly relevant detail: `omni_xpu_kernel.device.info(0)` classifies `0xE20B` as `bmg_sku=unknown` and `kernel_profile=generic-bmg`, while the docs describe generic BMG as the B70-compatible default. SYCL reports 5 slices x 4 subslices and `ocloc ids bmg-g21` maps to `20.1.0`, so the scheduler-unit query is nonzero and the device-image target appears consistent.

Questions:
1. Is Windows CUTE in this PR expected to support G21/B580, or only the validated G31/B70?
2. Should PCI ID `0xE20B` be classified with the G21/B60 policy?
3. Is oneAPI 2026.0 required, or should 2026.1 work?
4. Is there a smaller known-good forward probe or G21 tile/policy override you would like me to test?

I can provide a standalone minimal reproducer or test a targeted patch on this B580.

**Comment by allanmeng** [comment]
> Thanks for landing the Windows Torch 2.13 CUTE/Sol-Attn path. I tested PR #659 head `affc2ea6` on a physical Arc B580 (G21) and hit a reproducible forward hang before the real ComfyUI gate.
> 
> Environment:
> 
> * Windows, Arc B580, PCI ID `0xE20B`, `intel_gpu_bmg_g21`
> * driver `32.0.101.8974`
> * Python 3.13.12
> * PyTorch `2.13.0+xpu` (`torch.version.xpu == 20260000`)
> * oneAPI compiler/ocloc 2026.1; oneDNN 3.11.2
> * pinned sycl-tla `2fc09973bfdf15755090fcb0e3b6ad236408a992`
> 
> Because the PR's Windows CUTE target is hard-coded to `bmg-g31`, I made one isolated build-only change so `WINDOWS_CUTE_AOT_TARGETS["bmg"]` could be overridden, then built with `bmg-g21`. The wheel built and imported successfully:
> 
> * `omni_xpu_kernel 0.2.0b2+torch213.bmg`
> * SHA256 `4E35336F68E15FCCFB13A90B4F1FBFD9975DA6C946F9969A485F17994B7F68EE`
> * `cute.is_available()`, `supports_d128_bhld()`, `supports_minimax_h3_vae_d64()`, and `supports_sol_attn()` all returned true.
> 
> The matching PyTorch SDPA control completed with finite output, but each native CUTE forward remained in a CPU busy wait:
> 
> 1. H3 QKV-backed BF16 `[1,56,31,128]`, stride `[7168,128,21504,1]`: >124 s; repeat >34 s.
> 2. Dense BF16 BHLD `[1,4,31,128]`: >34 s.
> 3. Legacy FP16 BLHD `[1,256,8,128]`: >34 s.
> 
> There was no exception or explicit `DEVICE_LOST`; the Python process used about 1.4 GiB RAM and had to be terminated. Fresh Torch XPU matmul/device checks passed after each termination. I stopped before Sol-Attn forward and the real H3 `[1,56,7461,128]` gate.
> 
> One possibly relevant detail: `omni_xpu_kernel.device.info(0)` classifies `0xE20B` as `bmg_sku=unknown` and `kernel_profile=generic-bmg`, while the docs describe generic BMG as the B70-compatible default. SYCL reports 5 slices x 4 subslices and `ocloc ids bmg-g21` maps to `20.1.0`, so the scheduler-unit query is nonzero and the device-image target appears consistent.
> 
> Questions:
> 
> 1. Is Windows CUTE in this PR expected to support G21/B580, or only the validated G31/B70?
> 2. Should PCI ID `0xE20B` be classified with the G21/B60 policy?
> 3. Is oneAPI 2026.0 required, or should 2026.1 work?
> 4. Is there a smaller known-good forward probe or G21 tile/policy override you would like me to test?
> 
> I can provide a standalone minimal reproducer or test a targeted patch on this B580.

同款 0xE20B B580 验证：
- PR#659 原样 bmg-g31 AOT wheel（无任何 override）
- CUTE 前向正常：Qwen-Image D128 attention 实测命中
  （seq=4224, heads=24, bf16），无 hang
→ 建议用原样 g31 编译验证，g21 override 疑似 hang 根源

**Comment by allanmeng** [comment]
@yzwzhanghao2  补充一个同款卡的对照数据点。我的 Arc B580 也是 PCI 0xE20B
（device_id 57867，bmg_sku=unknown，kernel_profile=generic-bmg，跟你完全一致），
用 PR#659 原样 wheel（未做任何 override，AOT target 保持 bmg-g31）：

- 环境：Windows / torch 2.13.0+xpu / oneAPI 2026.1 / oneDNN 3.11.2 /
  sycl-tla pinned 2fc09973
- cute.is_available() 和三个 supports_* 全 True
- 实际 CUTE 前向无任何 hang：Qwen-Image GGUF DiT，d128 BHLD 自注意力
  （heads=24, seq=4224, bf16）每步命中 CUTE，稳态约 7.1 s/it，输出正常

所以同款 0xE20B 卡上，原样 bmg-g31 wheel 是健康的，而你 override 成
bmg-g21 编译的 wheel 全部 CUTE 前向卡死 → 疑似 g21 AOT image 在这张
卡上的问题，而非 Windows CUTE 通病。建议用原样 g31 复测一下，或
ocloc ids / sycl-ls 确认硬件实际绑定的 target。

-----------------------------------------------------------------------------
日志截图：
39a5d0319430b96a47c708ba3d399077

**Comment by yzwzhanghao2** [comment]
@allanmeng Thanks for the same-card data point. I retested with **unmodified PR #659 source** at `affc2ea6d15e93b5f2f68acb664fdf22c3506e8e`, keeping the default `bmg-g31` AOT target (no G21 override).

Environment: Arc B580 / PCI `0xE20B`, driver `32.0.101.8974`, Python 3.13.12, PyTorch `2.13.0+xpu` (XPU 20260000), oneAPI 2026.1, oneDNN 3.11.2. Wheel SHA256:

`A7896559FCD457D397AEF6F707E5A84FF8D6C75A49E4B2E97AA5C394529ED298`

CUTE extension SHA256:

`70B5999F573E704D640869FD3EF062A86C255B78C3B433AB601531FEC2048FE0`

Capabilities pass and ordinary BF16 XPU matmul remains healthy.

Results:

1. Packed BHLD `[1,24,4224,128]` BF16 completed once with finite output. Versus PyTorch SDPA: max abs `0.0009765625`, relative RMSE `0.0014657841`.
2. That process did not exit after writing the valid result; a second fresh run of the same shape did not finish within 45 s.
3. After reboot I matched the Qwen adapter layout: dense BLD `[1,4224,3072]` viewed as BHLD `[1,24,4224,128]`, stride `(12976128,128,3072,1)`. It also did not finish within 45 s.
4. The stuck process used ~1 CPU core, 1.46 GiB working set / 2.79 GiB private memory, and could not be terminated by either `Stop-Process -Force` or `taskkill /T /F`.
5. A separate XPU health probe still passed; no explicit `DEVICE_LOST`.

So G31 can execute one numerically valid forward here, but repeat/process-lifetime stability still differs from your B580. Matching Qwen stride was not sufficient.

Could you share your exact wheel/CUTE-extension SHA256, Intel driver and relevant SYCL DLL versions/search order, any `OMNI_*`/SYCL/Level Zero environment variables, and ideally a minimal probe doing 4 consecutive `sdp_bhld_d128` calls and exiting cleanly? This should isolate binary/runtime DLL selection versus queue/process initialization. I stopped before H3 D128, VideoVAE D64 and SolAttn to avoid more unkillable processes.

**Comment by allanmeng** [comment]
@yzwzhanghao2 Thanks for the detailed retest. Below is everything you asked for, plus a probe result that I think points at the root cause.

**Binaries (stock bmg-g31 AOT, no override, built from PR #659):**

- Wheel SHA256: `caa8fb32bd7e3ff44a6fc72475925c3dae69fa7aed0e1e1602844c6ca5d652f5`
- CUTE extension SHA256: `9138b7dafe8fe68a63cf8cac3ca4e54a8e1bbcc6f3278c20d8bfaa59445a6885`
- Intel driver: **32.0.101.8992** (yours is 8974 — note the delta)

**SYCL runtime:** torch-bundled DLLs in `python\Library\bin`: `sycl9.dll`, `sycl9-preview.dll`, `sycl-jit.dll`, plus `mkl_sycl_*.dll`. No standalone Level Zero loader there — Level Zero comes via the driver.

**Launcher env vars (these matter, see below):**

```
ZE_DEVICE_SLEEP=0
SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1
SYCL_PI_LEVEL_ZERO_TRACK_INDIRECT_ACCESS_MEMORY=1
SYCL_PI_LEVEL_ZERO_DEVICE_SCOPE_EVENTS=0
ONEAPI_DEVICE_SELECTOR=level_zero:0
SYCL_DEVICE_FILTER=level_zero
ZE_ENABLE_PCI_ID_DEVICE_ORDER=1
PYTORCH_XPU_ALLOC_CONF=max_split_size_mb:64
OMNIXPU_ENABLE=1 / OMNI_ATTN_BACKEND=cute / SOL_ATTN_XPU_EXPERIMENTAL=1
```
Plus oneAPI `setvars.bat` on PATH (compiler/tbb/dnnl dirs ahead of system).

**Minimal probe (4 consecutive `sdp_bhld_d128`, Qwen-like `[1,24,4224,128]` bf16):**

```
call #1: 0.009s finite=True
call #2: 0.003s finite=True
call #3: 0.003s finite=True
call #4: 0.003s finite=True
stride-view call: 0.004s finite=True
ALL DONE - clean exit (exit 0)
```

**Key observation — the probe only passes WITH the Level Zero env vars above.** Running the exact same probe in a bare environment (no `ZE_DEVICE_SLEEP`, no `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS`, etc.) the process hangs at init and never reaches the kernel. With them set, all four consecutive calls complete in single-digit milliseconds and the process exits cleanly.

So before blaming the g31 AOT image: could you retry your standalone reproducer with at least
`ZE_DEVICE_SLEEP=0` and `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1` set? Your symptom (first call ok, second/process-lifetime hang, unkillable process) matches the device-sleep / command-list path that these two vars control on BMG — I suspect the missing env is the actual difference between your runs and mine, not the binary.

Also worth a try: update the driver from 8974 to 8992+ if you can.

**minimal: 4x sdp_bhld_d128 then clean exit (run WITH the env vars above)**
import torch, time
torch.xpu.init()
from omni_xpu_kernel import cute
q = torch.randn(1, 24, 4224, 128, dtype=torch.bfloat16, device='xpu')
k, v = q.clone(), q.clone()
for i in range(4):
    t = time.perf_counter()
    out = cute.sdp_bhld_d128(q, k, v)
    torch.xpu.synchronize()
    print(f'call {i+1}: {time.perf_counter()-t:.3f}s finite={bool(torch.isfinite(out).all())}')
print('clean exit')
